tests/test_interpolation.py

```diff
@@ -753,3 +758,251 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
+                        """
+                    )
+                ),
+            ),
+            id="type_mismatch_node_interpolation",
+        ),
+        pytest.param(
+            StructuredWithMissing(opt_num=None, num=II("opt_num")),
+            "num",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Non optional field cannot be assigned None"),
+            ),
+            id="non_optional_node_interpolation",
+        ),
+        pytest.param(
+            SubscriptedList(list=SI("${identity:[a, b]}")),
+            "list",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Value 'a' could not be converted to Integer"),
+            ),
+            id="list_type_mismatch",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{0: b, 1: d}}")),
+            "dict",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Key 0 (int) is incompatible with (str)"),
+            ),
+            id="dict_key_type_mismatch",
+        ),
+    ],
+)
+def test_interpolation_type_validated_error(
+    cfg: Any,
+    key: str,
+    expected_error: Any,
+    restore_resolvers: Any,
+) -> None:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    OmegaConf.register_new_resolver("cast", cast)
+    OmegaConf.register_new_resolver("identity", lambda x: x)
+
+    cfg = OmegaConf.structured(cfg)
+
+    with expected_error:
+        cfg[key]
+
+    assert OmegaConf.select(cfg, key, throw_on_resolution_failure=False) is None
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key"),
+    [
+        pytest.param({"dict": "${identity:{a: 0, b: 1}}"}, "dict.a", id="dict"),
+        pytest.param(
+            {"dict": "${identity:{a: 0, b: {c: 1}}}"},
+            "dict.b.c",
+            id="dict_nested",
+        ),
+        pytest.param({"list": "${identity:[0, 1]}"}, "list.0", id="list"),
+        pytest.param({"list": "${identity:[0, [1, 2]]}"}, "list.1.1", id="list_nested"),
+    ],
+)
+def test_interpolation_readonly_resolver_output(
+    restore_resolvers: Any, cfg: Any, key: str
+) -> None:
+    OmegaConf.register_new_resolver("identity", lambda x: x)
+    cfg = OmegaConf.create(cfg)
+    sub_key: Any
+    parent_key, sub_key = key.rsplit(".", 1)
+    try:
+        sub_key = int(sub_key)  # convert list index to integer
+    except ValueError:
+        pass
+    parent_node = OmegaConf.select(cfg, parent_key)
+    with pytest.raises(ReadonlyConfigError):
+        parent_node[sub_key] = -1
+
+
+def test_interpolation_readonly_node() -> None:
+    cfg = OmegaConf.structured(User(name="7", age=II("name")))
+    resolved = cfg._get_node("age")._dereference_node()
+    assert resolved == 7
+    with pytest.raises(ReadonlyConfigError):
+        resolved._set_value(8)
```

Comment:
I thought we decided to make generated (transient) nodes readonly, not real nodes.

Is this leaking out to users?

Reply:
Not sure what you mean, what wouldn't be a "real node"?

This specific test is using internal functions to test it on non-container nodes. I can't think of a way users using the public API can run into this, but I still wanted to test it for completeness.

Reply:
This is an interpolation from one node to another.
We have a use case we want to support:
```python
cfg = OmegaConf.create({
 "user":  {"age" : 7},
 "actor" : "${user}"
})

assert cfg.actor.age == 7
cfg.actor.age = 8
assert cfg.user.age = 8
```

This means that returned container nodes from interpolations are not read only.
Somehow, in this case the resolved value node is read-only.
Is it read-only because the validation generated a new node?

Reply:
> Is it read-only because the validation generated a new node?
Yes it's only read-only for nodes that are generated on-the-fly.

Also, the example you gave is for a container node (`user`): when we write `cfg.user.age = 8` we are actually writing (first) into `cfg.user`, not `cfg.user.age`. This means that if `cfg.user` is read-only, then this assignment will fail (even if we explicitly set the `"readonly"` flag to False on the `age` node).

Reply:
1. The example is wrong, the write is through actor. The assertion should be `assert cfg.user.age == 8`.
2. cfg.user is not read-only.
3. Assignment of into DictConfig is complicated. :)

Having said all that, I think we can can resolve this.

Reply:
Regarding point 2, yes, sorry that was a bit confusing. I wanted to emphasize that your example was testing the assignment into a DictConfig returned by an interpolation, and not into a ValueNode (while this test is specifically to test assignment into a ValueNode).

I can't think of a situation where we would try to assign into ValueNode generated by interpolation resolution (using the public API).

Feel free to resolve :)